enhancement(kubernetes_logs): preserve event ordering

sillent opened the pull request:
PR that may solve the problem in issue #10883

sillent committed:
implement new optional field order for kubernetes_logs sources

the field is responsible for the logic of the order in which events are processed

bruceg commented on src/sources/kubernetes_logs/mod.rs:
```diff
@@ -445,10 +454,13 @@ impl Source {
-        let mut lifecycle = Lifecycle::new();
+        let mut lc: Lifecycle;
+        match ordered {
+            true => lc = Lifecycle::new(lifecycle::LifecycleOrder::Ordered),
+            false => lc = Lifecycle::new(lifecycle::LifecycleOrder::Unordered),
```
If you move the `ordered` parameter to `Lifecycle::new`, I think you can hide the `LifecycleOrder` enum entirely in that module and not have it exposed here. I would also recommend against shortening the name.

sillent replied:
@bruceg now into Lifecycle::new just passes an 'ordered' as arguments

bruceg reviewed (changes requested):
Thank you for your contribution, @sillent. It's not obvious to me that this approach will solve the problem in #10883, but I am not very familiar with the `kubernetes_logs` source so I'll leave that commentary to someone else. I did leave a couple of requests for improving the code patterns. Also, make sure to format your code before submitting by running `cargo fmt` if your editor doesn't do that automatically.

sillent committed:
fix website documentation - triple quote added

spencergilbert commented on src/sources/kubernetes_logs/mod.rs:
```diff
@@ -138,6 +138,10 @@ pub struct Config {
     /// event from the watched stream.
     delay_deletion_ms: usize,
+
+    /// What's order the events from log files must be passed to the sinks
+    /// By default all events is unordered
```
```suggestion
    /// If events must be properly ordered when sending downstream,
    /// by default all events are unordered.
```

sillent replied:
fixed

spencergilbert commented on website/cue/reference/components/sources/kubernetes_logs.cue:
```diff
@@ -288,6 +288,13 @@ components: sources: kubernetes_logs: {
+		order: {
+			common: false
+			description: """
+				Specifies the order in which events are processed.
```
```suggestion
				Specifies if the events are processed in order.
```

sillent replied:
fixed, and closed quotes added

spencergilbert commented on src/sources/kubernetes_logs/lifecycle.rs:
```diff
@@ -118,7 +164,11 @@ impl<'bound, 'lc> Slot<'bound, 'lc> {
+            LifecycleOrder::Ordered => self.lifecycle.futs_ordered.push(future),
+            LifecycleOrder::Unordered => self.lifecycle.futs_unordered.push(future),
+        }
+        // self.lifecycle.futs.push(future);
```
```suggestion
```

Old code left commented out

sillent replied:
fixed

spencergilbert reviewed (commented):
I left a couple of nits around docs, comments, and a stray old line. Similarly to @bruceg and @jszwedko, it's unclear to me how the lifecycle code is influencing the order in which events are processed by the underlying file server.

Given that with the patch, you've (manually?) tested and confirmed the behavior is corrected, I'm curious if we can include some sort of test that confirms, and prevents regressions, in the ordered processing of events here.

As an aside, I'm also wondering if this option would be better suited in the underlying file server that processes the log files - though I know we do plan on rewriting much of that component.

sillent committed:
refactoring by comments in PR

- removed an enum
- change logic by the passing bool value in Lifecycle
- some changes from the PR comments

sillent commented on src/sources/kubernetes_logs/lifecycle.rs:
```diff
@@ -8,12 +8,18 @@ use futures::{
+pub enum LifecycleOrder{
+    Ordered,
+    Unordered,
+}
```
@bruceg thank you for comments. I've changed the logic. Now the ordering is passed to the new method of Lifecycle.

sillent commented:
> I left a couple of nits around docs, comments, and a stray old line. Similarly to @bruceg and @jszwedko, it's unclear to me how the lifecycle code is influencing the order in which events are processed by the underlying file server.
> 
> Given that with the patch, you've (manually?) tested and confirmed the behavior is corrected, I'm curious if we can include some sort of test that confirms, and prevents regressions, in the ordered processing of events here.
> 
> As an aside, I'm also wondering if this option would be better suited in the underlying file server that processes the log files - though I know we do plan on rewriting much of that component.

@spencergilbert hello, thank you for you comment

I'm not sure but I think it's a problem with the order of events is starting here
`
  let mut lifecycle = Lifecycle::new();
        {
            let (slot, shutdown) = lifecycle.add();
            let fut = util::run_file_server(file_server, file_source_tx, shutdown, checkpointer)
                .map(|result| match result {
                    Ok(FileServerShutdown) => info!(message = "File server completed gracefully."),
                    Err(error) => emit!(KubernetesLifecycleError {
                        message: "File server exited with an error.",
                        error,
                    }),
                });
            slot.bind(Box::pin(fut));
        }
`
when the start a FileServer, that watching for new files and start reading it and generate new events.

The problem with FuturesUnordered is appear when the file fill very fast (application write log too fast), when the first file is filled up, and system rotating him, the Vector is still not fully processed him, and the new files appears, the 'FileServer' start reading the new log-file and start generating events. The events from first file is mixing up by the events from the new file, and in destination we saw a unordered lines.

bruceg commented:
I think your explanation makes sense. Could you confirm that using a `FuturesOrdered` resolves the problem you are seeing. If so, is there any reason not to use it unconditionally?

spencergilbert commented:
> I think your explanation makes sense. Could you confirm that using a `FuturesOrdered` resolves the problem you are seeing. If so, is there any reason not to use it unconditionally?

Yeah, that's reasonable - unless there's a noticeable performance impact it seems like this would be strictly "better" behavior.

sillent commented:
> Yeah, that's reasonable - unless there's a noticeable performance impact it seems like this would be strictly "better" behavior.

In issue #10883 my colleague was testing the functionality with FuturesOrdered and FuturesUnordered, and with FuturesOrdered the message order is not violated, however with the same number of message with FuturesUnordered we have order breaking.
So if the FuturesOrdered is preferred for this situation, i'l do changed the FuturesUnordered to FuturesOrdered and is it enough?

spencergilbert commented:
Thanks again @sillent - it sounds like just switching to `FuturesOrdered` is the best way forward, given it seems like a strictly better default.

sillent committed:
only changed FuturesUnordered to FuturesOrdered

sillent commented:
> Thanks again @sillent - it sounds like just switching to `FuturesOrdered` is the best way forward, given it seems like a strictly better default.

Cool, i switched back the all changes and do only one  - change FuturesUnordered to FuturesOrdered

bruceg committed:
Merge remote-tracking branch 'origin/master' into 10883-fix-order

bruceg reviewed (approved):
All checks are green and the change seems simple enough. I'm good with this if others are.

spencergilbert commented:
Thanks again for the contribution @sillent!
